src/main/scala/org/apache/spark/mllib/optimization/tfocs/SmoothFunction.scala

```diff
@@ -17,58 +17,164 @@
+      // Compute the squared error.
+      Some(g.treeAggregate(0.0)((sum, y) => sum + Math.pow(Vectors.norm(y, 2), 2), _ + _) / 2.0)
+    } else {
+      None
+    }
     Value(f, Some(g))
   }
 }
 
-/** The squared error function applied to RDD[Vector] vectors. */
-class SquaredErrorRDDVector(x0: RDD[Vector]) extends SmoothFunction[RDD[Vector]] {
+/**
+ * The huber loss function applied to a DVector.
+ *
+ * @param x0 The vector against which loss should be computed.
+ * @param tau The huber loss parameter.
+ *
+ * NOTE In matlab tfocs this functionality is implemented in smooth_huber.m.
+ */
+class SmoothHuber(x0: DVector, tau: Double) extends SmoothFunction[DVector] with Serializable {
 
   x0.cache()
 
-  override def apply(x: RDD[Vector], mode: Mode): Value[Double, RDD[Vector]] = {
-    val g = x.zip(x0).map(y =>
-      new DenseVector(y._1.toArray.zip(y._2.toArray).map(z => z._1 - z._2)): Vector)
-    if (mode.f && mode.g) g.cache()
+  override def apply(x: DVector, mode: Mode): Value[DVector] = {
+
+    val diff = x.diff(x0)
+
+    // If both f and g are requested then diff will be read twice, so cache it.
+    if (mode.f && mode.g) diff.cache()
+
     val f = if (mode.f) {
-      Some(g.treeAggregate(0.0)((sum, x) => sum + Math.pow(Vectors.norm(x, 2), 2), _ + _) / 2.0)
+      Some(diff.aggregateElements(0.0)(
+        seqOp = (sum, diff_i) => {
+          // Find the huber loss, corresponding to the adjusted l2 loss when the magnitude is <= tau
+          // and to the adjusted l1 loss when the magnitude is > tau.
+          val huberValue = if (math.abs(diff_i) <= tau) {
+            0.5 * diff_i * diff_i / tau
+          } else {
+            math.abs(diff_i) - tau / 2.0
+          }
+          sum + huberValue
+        },
+        combOp = _ + _))
     } else {
       None
     }
-    Value(f, Some(g))
+
+    val g = if (mode.g) {
+      // Compute the huber loss gradient elementwise.
+      Some(diff.mapElements(diff_i => diff_i / math.max(math.abs(diff_i), tau)))
+    } else {
+      None
+    }
+
+    Value(f, g)
   }
 }
 
+/**
+ * The log likelihood logistic loss function applied to a DVector.
+ *
+ * @param y The observed values, labeled as binary 0/1.
+ * @param mu The variable values.
+ *
+ * NOTE In matlab tfocs this functionality is implemented in smooth_logLLogistic.m.
+ */
+class SmoothLogLLogistic(y: DVector) extends SmoothFunction[DVector] with Serializable {
+
+  y.cache()
+
+  override def apply(mu: DVector, mode: Mode): Value[DVector] = {
+
+    val f = if (mode.f) {
+      Some(y.zip(mu).treeAggregate(0.0)(
+        seqOp = (_, _) match {
+          case (sum, (yPart, muPart)) => {
+
+            // Check that the y and mu partitions are the same size.
+            if (yPart.size != muPart.size) {
+              throw new IllegalArgumentException("Can only zip Vectors with the same number of " +
+                "elements")
+            }
+
+            yPart.toArray.zip(muPart.toArray).aggregate(0.0)(
```

Comment:
Ditto. `zip` hurts the performance. We should use `while` loop instead. If we don't want to spend too much time on it, we can use `toArray.view.zip` instead of `toArray.zip`. It might help, but I'm not sure.